shell/platform/android/io/flutter/embedding/android/KeyEmbedderResponder.java

```diff
@@ -341,12 +342,33 @@ private boolean handleEventImpl(
     }
 
     final KeyData output = new KeyData();
+
+    switch (event.getSource()) {
+      default:
+      case InputDevice.SOURCE_KEYBOARD:
+        output.deviceType = KeyData.DeviceType.kKeyboard;
+        break;
+      case InputDevice.SOURCE_DPAD:
+        output.deviceType = KeyData.DeviceType.kDirectionalPad;
+        break;
+      case InputDevice.SOURCE_GAMEPAD:
+        output.deviceType = KeyData.DeviceType.kGamepad;
+        break;
+      case InputDevice.SOURCE_JOYSTICK:
+        output.deviceType = KeyData.DeviceType.kJoystick;
+        break;
+      case InputDevice.SOURCE_HDMI:
+        output.deviceType = KeyData.DeviceType.kHdmi;
+        break;
+    }
+
     output.timestamp = event.getEventTime();
     output.type = type;
     output.logicalKey = logicalKey;
     output.physicalKey = physicalKey;
     output.character = character;
     output.synthesized = false;
+    output.deviceType = KeyData.DeviceType.kKeyboard;
```

Comment:
From Android triage (specifically https://github.com/flutter/flutter/issues/151308):

Does this line override the above switch statement in all cases?

Reply:
Yeah, seems like it does.  That line should probably be removed.